Make `ParameterTable` use rust-native errors.

raynelfss opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary
- Add one new error variant, `ConflictingName`, to `ParameterTableError`, to account for instances with conflicting names.
- Add type `ParameterResult<T>` as a shorthand result handler for `ParameterTableError`.
- Added variant `ConflictingName` to condition in `CircuitData::set_global_phase`.

### Details and comments
- Prelude to #15027

raynelfss committed:
Initial: Make `ParameterTable` use rust-native errors.
- Add one new error variant, `ConflictingName`, to `ParameterTableError`, to account for instances with conflicting names.
- Add type `ParameterResult<T>` as a shorthand result handler for `ParameterTableError`.
- Added variant `ConflictingName` to condition in `CircuitData::set_global_phase`.

jakelishman commented on crates/circuit/src/parameter_table.rs:
```diff
@@ -26,12 +26,16 @@ use crate::parameter::symbol_expr::Symbol;
 
 import_exception!(qiskit.circuit, CircuitError);
 
+pub type ParameterTableResult<T> = Result<T, ParameterTableError>;
```
Please can we avoid this if at all possible - in my experience of dealing with these, it adds more mental load in trying to remember what the wrapper class means, and makes it harder to jump to the error type.  It can make sense as an alias when you're not _supposed_ to be able to match on the error types, but if you are (and that's entirely reasonable here), then hiding the type makes it harder.

raynelfss replied:
Oh, I didn't think it would affect things badly since it's just a name alias. I could even keep the type private just for use within the file itself and expose it as `Result<T, ParameterTableError>` to everything else. I can also remove it if it seems cumbersome.

jakelishman commented on crates/circuit/src/parameter_table.rs:
```diff
@@ -26,12 +26,16 @@ use crate::parameter::symbol_expr::Symbol;
+
 #[derive(Error, Debug)]
 pub enum ParameterTableError {
     #[error("parameter '{0:?}' is not tracked in the table")]
     ParameterNotTracked(ParameterUuid),
     #[error("usage {0:?} is not tracked by the table")]
     UsageNotTracked(ParameterUse),
+    #[error("name conflict adding parameter '{0}'")]
+    ConflictingName(String), // CircuitError
 }
```
I'm not convinced that adding `ConflictingName` here actually fits.  `ConflictingName` is a _user_ error, the other two are internal logic errors - it should be literally impossible to trigger them from Python space.  Honestly, the other two probably could be straight panics (since they're internal logic bugs) and the error class might not have needed to exist at all previously, but it's relatively harmless.

On the other hand, `ConflictingName` represents a _user_ error of the circuit.  In interactive programs, it definitely shouldn't be fatal to the process - it's something we should report to the user so they can fix it.  So the handling for `ConflictingName` is always going to be different to the handling for the other two, and `ConflictingName` arises from different methods to all the others.

I think it might well be easier to leave `ParameterTableError` as it is and instead return `Result<ParameterUuid, ParameterUuid>` from `track` - the `Ok` variant holds the id of the good symbol, and the `Err` variant holds the uuid of the existing, conflicting variable (since the `name` is already accessible to the caller through the shared `&Symbol` - they don't give up their ownership).

Cryoris replied:
I'm curious how this looks when we remove `PyResult` up the chain and have `CircuitData::from_packed_instruction` return a Rust-specific error. From C we could also have conflicting names, so these needs to be propagated up somehow, right?

raynelfss committed:
FIx: Modify result in `ParameterTable.track`
- Based on Jake's suggestion, use `ParameterUuid` as the error type to both mark that an error happened and to specify the uuid of the already existent `Parameter`.
- Remove `ParameterTableResult` alias.
- Added function `track_parameter` to return `CircuitError` when operating in `CircuitData`.

eliarbel commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -2985,6 +2991,22 @@ impl<'py> FromPyObject<'py> for AssignParam {
 
+/// Wrapper around [ParameterTable::track] that returns a [CircuitError] if the parameter
+/// already exists.
+fn track_parameter(
```
Any special reason why this is not a `CircuitData` method?

raynelfss replied:
Yes, it is mainly because of the method `track_instruction_parameters`. It was a way of getting around borrowing the circuit immutably to retrieve an instruction and iterating through its parameters, and then mutably during the iteration to track the parameter. Getting around it by just sending a mutable reference to the parameter table gets around that, but I will see if I can fix this in a cleaner way.

raynelfss commented:
Marked for closure by #15279
